File: src/main/java/no/borber/monsterShop/application/basket/BasketApplicationService.java
package no.borber.monsterShop.application.basket;

import no.borber.monsterShop.application.AggregateType;
import no.borber.monsterShop.application.order.OrderAggregate;
import no.borber.monsterShop.eventStore.EventStore;
import no.borber.monsterShop.eventStore.Event;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

public class BasketApplicationService {

    private EventStore eventStore;

    public BasketApplicationService(EventStore eventStore) {
        this.eventStore = eventStore;
    }

    public void createBasket(String id) {
        BasketAggregate basketAggregate = new BasketAggregate(eventStore.getById(AggregateType.BASKET, id));
        basketAggregate.createBasket(id);
        List<Event> derivedEvents = basketAggregate.getDerivedEvents();
        eventStore.store(derivedEvents);
    }

    public void addItemToBasket(String id, String monsterType) {
        BasketAggregate basketAggregate = new BasketAggregate(eventStore.getById(AggregateType.BASKET, id));
        basketAggregate.addItemToBasket(monsterType);
        List<Event> derivedEvents = basketAggregate.getDerivedEvents();
        eventStore.store(derivedEvents);
    }

    public void removeItemFromBasket(String id, String monsterType) {
        BasketAggregate basketAggregate = new BasketAggregate(eventStore.getById(AggregateType.BASKET, id));
        basketAggregate.removeItemFromBasket(monsterType);
        List<Event> derivedEvents = basketAggregate.getDerivedEvents();
        eventStore.store(derivedEvents);
    }

    public void checkoutBasket(String basketId, String customerId, String orderId) {
        BasketAggregate basket = new BasketAggregate(eventStore.getById(AggregateType.BASKET, basketId));
        List<BasketLineItem> currentBasketLineItems = basket.checkoutBasket();

        OrderAggregate order = new OrderAggregate(eventStore.getById(AggregateType.ORDER, orderId));
        order.createOrder(orderId, customerId, currentBasketLineItems, LocalDateTime.now());

        eventStore.store(basket.getDerivedEvents());
        eventStore.store(order.getDerivedEvents());
    }
}

File: src/main/scala/no/borber/monsterShop/application/basket/BasketApplicationService.scala
package no.borber.monsterShop.application.basket

import akka.actor.{ActorRef, ActorSystem}
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

import scala.collection.mutable


@Component
class BasketApplicationService {

  @Autowired
  private var system: ActorSystem = null

  var  baskets = new mutable.HashMap[String, ActorRef];


  def this(system: ActorSystem) {
    this()
    this.system = system
  }

  def createBasket(id: String) {
    getBasketActor(id) ! CreateBasket
  }

  def getBasketActor(id: String): ActorRef = {
    baskets.getOrElseUpdate(id,system.actorOf(BasketActor.props("basket-"+id)));
  }

  def addItemToBasket(id: String, monsterType: String) {
    getBasketActor(id) ! AddItemToBasket(monsterType)
  }

  def removeItemFromBasket(id: String, monsterType: String) {
    getBasketActor(id) ! RemoveItemFromBasket(monsterType)
  }

  def checkoutBasket(basketId: String, customerId: String, orderId: String) {
    /*val basket: BasketAggregate = new BasketAggregate(eventStore.getById(basketId))
    val currentBasketLineItems: List[BasketLineItem] = basket.checkoutBasket
    val order: OrderAggregate = new OrderAggregate(eventStore.getById(orderId))
    order.createOrder(orderId, customerId, currentBasketLineItems, LocalDateTime.now)
    val derivedEvents: List[Event] = new ArrayList[Event]
    derivedEvents.addAll(basket.getDerivedEvents)
    derivedEvents.addAll(order.getDerivedEvents)
    eventStore.store(derivedEvents)*/
  }
}
